### DISCUSSION TITLE
did anyone know how Remix handle the form resubmission problem?

### DISCUSSION THREAD

**Discussion by lili21** [remix-run/remix#6786] [Q&A]
when the JS is not available for some reason, and the form submission fall back to native browser behavior, how does remix prevent the form resubmission problem?

**Comment by kiliman** [comment]
The typical technique is `POST/REDIRECT/GET` (PRG).

You should redirect from your action. This way if the user tries to refresh, the browser will not resubmit the form.

**Reply by lili21** [reply]
I discovered that the remix already addressed the issue without requiring the use of PRG. I'm curious to know how the remix achieved it.

**Reply by lili21** [reply]
I figured it out. it was the `<ScrollRestoration />` component. it will call the `window.history.replaceState` function, which addressed the issue.
